File: lib/domain/services/sound_detection.dart
import '../models/recording.dart';

const yamnetSampleRate = 16000;
const yamnetWindowSamples = 15600;
const yamnetWindowSeconds = yamnetWindowSamples / yamnetSampleRate;

const _speechIndex = 0;
const _snoringIndex = 38;
const _minimumConfidence = 0.32;
const _mergeGapSeconds = 1.5;

class SoundClassification {
  const SoundClassification({required this.kind, required this.confidence});

  final SoundEventKind kind;
  final double confidence;
}

SoundClassification? classifyYamnetScores(List<double> scores) {
  final speechScore = scores.length > _speechIndex ? scores[_speechIndex] : 0.0;
  final snoringScore = scores.length > _snoringIndex
      ? scores[_snoringIndex]
      : 0.0;

  if (speechScore < _minimumConfidence && snoringScore < _minimumConfidence) {
    return null;
  }

  if (snoringScore >= speechScore) {
    return SoundClassification(
      kind: SoundEventKind.snoring,
      confidence: snoringScore,
    );
  }

  return SoundClassification(
    kind: SoundEventKind.speech,
    confidence: speechScore,
  );
}

List<SoundEvent> appendSoundEvent(
  List<SoundEvent> events,
  SoundClassification classification,
  double startSeconds,
) {
  final endSeconds = startSeconds + yamnetWindowSeconds;
  final previous = events.isEmpty ? null : events.last;

  if (previous != null &&
      previous.kind == classification.kind &&
      startSeconds <= previous.endSeconds + _mergeGapSeconds) {
    return [
      ...events.take(events.length - 1),
      previous.copyWith(
        endSeconds: endSeconds > previous.endSeconds
            ? endSeconds
            : previous.endSeconds,
        confidence: classification.confidence > previous.confidence
            ? classification.confidence
            : previous.confidence,
      ),
    ];
  }

  return [
    ...events,
    SoundEvent(
      id: '${classification.kind.jsonValue}-${(startSeconds * 1000).round()}',
      kind: classification.kind,
      startSeconds: startSeconds,
      endSeconds: endSeconds,
      confidence: classification.confidence.clamp(0, 1).toDouble(),
    ),
  ];
}

List<double>? createYamnetFrame(Iterable<double> samples) {
  final source = samples is List<double>
      ? samples
      : samples.toList(growable: false);
  if (source.length < yamnetWindowSamples ~/ 4) return null;

  final frame = List<double>.filled(yamnetWindowSamples, 0);
  final copyLength = source.length < yamnetWindowSamples
      ? source.length
      : yamnetWindowSamples;
  for (var index = 0; index < copyLength; index += 1) {
    frame[index] = source[index];
  }
  return frame;
}

List<double> resampleLinear(List<double> samples, int sourceSampleRate) {
  if (samples.isEmpty ||
      sourceSampleRate <= 0 ||
      sourceSampleRate == yamnetSampleRate) {
    return samples;
  }

  final ratio = yamnetSampleRate / sourceSampleRate;
  final outputLength = (samples.length * ratio).round().clamp(
    1,
    samples.length * 4,
  );
  final resampled = List<double>.filled(outputLength, 0);
  for (var index = 0; index < outputLength; index += 1) {
    final sourceIndex = index / ratio;
    final leftIndex = sourceIndex.floor().clamp(0, samples.length - 1);
    final rightIndex = (leftIndex + 1).clamp(0, samples.length - 1);
    final mix = sourceIndex - leftIndex;
    resampled[index] =
        samples[leftIndex] * (1 - mix) + samples[rightIndex] * mix;
  }
  return resampled;
}

File: src/features/recordings/sound-detection.ts
import type { SoundEvent, SoundEventKind } from './recording-types';

export const YAMNET_SAMPLE_RATE = 16_000;
export const YAMNET_WINDOW_SAMPLES = 15_600;
export const YAMNET_WINDOW_SECONDS = YAMNET_WINDOW_SAMPLES / YAMNET_SAMPLE_RATE;

const SPEECH_INDEX = 0;
const SNORING_INDEX = 38;
const MIN_CONFIDENCE = 0.32;
const MERGE_GAP_SECONDS = 1.5;

export interface SoundClassification {
  kind: SoundEventKind;
  confidence: number;
}

export function classifyYamnetScores(scores: Float32Array): SoundClassification | null {
  const speechScore = scores[SPEECH_INDEX] ?? 0;
  const snoringScore = scores[SNORING_INDEX] ?? 0;

  if (speechScore < MIN_CONFIDENCE && snoringScore < MIN_CONFIDENCE) {
    return null;
  }

  if (snoringScore >= speechScore) {
    return { kind: 'snoring', confidence: snoringScore };
  }

  return { kind: 'speech', confidence: speechScore };
}

export function appendSoundEvent(
  events: SoundEvent[],
  classification: SoundClassification,
  startSeconds: number,
): SoundEvent[] {
  const endSeconds = startSeconds + YAMNET_WINDOW_SECONDS;
  const previous = events.at(-1);

  if (
    previous &&
    previous.kind === classification.kind &&
    startSeconds <= previous.endSeconds + MERGE_GAP_SECONDS
  ) {
    return [
      ...events.slice(0, -1),
      {
        ...previous,
        endSeconds: Math.max(previous.endSeconds, endSeconds),
        confidence: Math.max(previous.confidence, classification.confidence),
      },
    ];
  }

  return [
    ...events,
    {
      id: `${classification.kind}-${Math.round(startSeconds * 1000)}`,
      kind: classification.kind,
      startSeconds,
      endSeconds,
      confidence: classification.confidence,
    },
  ];
}

export function createYamnetFrame(samples: number[]): Float32Array | null {
  if (samples.length < YAMNET_WINDOW_SAMPLES / 4) {
    return null;
  }

  const frame = new Float32Array(YAMNET_WINDOW_SAMPLES);
  const copyLength = Math.min(samples.length, YAMNET_WINDOW_SAMPLES);
  frame.set(samples.slice(0, copyLength));
  return frame;
}
